WIP: ARROW-7871: [Python] [DOC] Expose more compute kernels and add compute module docs

arw2019 opened the pull request.

arw2019 commented on docs/source/cpp/compute.rst:
```diff
@@ -201,6 +201,11 @@ an ``Invalid`` :class:`Status` when overflow is detected.
 +--------------------------+------------+--------------------+---------------------+
 | subtract_checked         | Binary     | Numeric            | Numeric             |
 +--------------------------+------------+--------------------+---------------------+
+| divide                   | Binary     | Numeric            | Numeric             |
```
Noticed that this was missing on the c++ page

pitrou replied:
Ah, thank you! However, can you leave the table alphabetically-sorted?

arw2019 replied:
Actually my bad, you already had it there. Reverted

jorisvandenbossche commented:
@arw2019 thanks a lot for working on this! (and sorry for the overlap with https://github.com/apache/arrow/pull/8163)

jorisvandenbossche commented on docs/source/python/api/compute.rst:
```diff
@@ -0,0 +1,206 @@
+
+   ascii_lower
+   ascii_upper
+   binary_length
```
this is not really a transform?

arw2019 replied:
Yeah that's fair. Would `Structural Transforms` be a good place for it?

jorisvandenbossche commented on docs/source/python/compute.rst:
```diff
@@ -0,0 +1,54 @@
+Compute Functions
+=================
+
+Arrow supports logical compute operations over inputs of possibly 
```
what does the "logical" in here mean exactly?

arw2019 replied:
So I actually lifted the meat of this section from the corresponding C++ doc:
https://arrow.apache.org/docs/cpp/compute.html
Do we want to just delete `logical` here and in the C++ page?

jorisvandenbossche replied:
cc @pitrou

pitrou replied:
It was meant to convey that the operations are not typed a priori (you have a single "add" function for multiple possible types, for example). But perhaps that is superfluous.

jorisvandenbossche commented on docs/source/python/compute.rst:
```diff
@@ -0,0 +1,54 @@
+   >>> a = pa.array([1, 1, 2, 3])
+   >>> pc.sum(a)
+   <pyarrow.Int64Scalar: 7>
+   >>> b = pa.array([4, 5, 8, 2])
```
this array is not used? (maybe show pc.equals(a, b) ?)

arw2019 commented on python/pyarrow/compute.py:
```diff
@@ -294,6 +294,46 @@ def mode(array):
     return call_function("mode", [array])
 
 
+def and_(x1, x2):
```
@jorisvandenbossche @pitrou 

AFAIK we can't have methods named `and` and `or` in Python. For now I went with `and_` and `or_` here, for consistency with `and_kleene` and `or_kleene`, but for example numpy uses `np.logical_and` and `np.logical_or`

pitrou replied:
We already have a function named "and":
```python
>>> from pyarrow import compute
>>> getattr(compute, "and")
<function pyarrow.compute.and(left, right, *, memory_pool=None)>
```
So you should just fix `compute.py` so that it's exported as `and_`.

arw2019 replied:
Right. Fixed now

pitrou commented on docs/source/python/api/compute.rst:
```diff
@@ -0,0 +1,206 @@
+   mean
+   min_max
+   sum
+   mode
```
Will this make a link to the API docs?

arw2019 replied:
I *think* so. It's what's done on the other pages:
https://github.com/apache/arrow/blob/master/docs/source/python/api/datatypes.rst

I'm having trouble building the docs locally (related to ARROW-10018 I think) so haven't been able to check

jorisvandenbossche replied:
This won't render the docstring here, but it will create a separate page for each of those functions and link to that from this table. 

(the separate pages might be a bit overkill since the docstrings here are often not that informative yet, but it's indeed how we do it for other functions as well)

arw2019 commented:
Question: in 952649d I'm attempting to expose `PartitionNthOptions` but it errors at compilation. What is (are) my mistake(s)?

This is needed to make the `partition_nth_indices` Python wrapper usable.

pitrou commented:
Cython generally doesn't convert implicitly between arbitrary Python objects and C/C++ types, you have to type your code explicitly. Also, it is [better](https://cython.readthedocs.io/en/latest/src/userguide/special_methods.html#initialisation-methods-cinit-and-init) to use `__cinit__` here so that the constructor never gets skipped, so for example (untested):
```cython
    def __cinit__(self, int64_t pivot):
        self.partition_nth_options.reset(new CPartitionNthOptions(pivot))
```

arw2019 commented:
> Cython generally doesn't convert implicitly between arbitrary Python objects and C/C++ types, you have to type your code explicitly. 

Thanks @pitrou! That was the problem

arw2019 commented:
This is ready for re-review. 

I believe that I've addressed the feedback from previous reviews. I've also now exposed all the option classes so that all the kernels listed in the docs are usable from Python.

jorisvandenbossche commented on docs/source/python/compute.rst:
```diff
@@ -0,0 +1,55 @@
+
+.. seealso::
+   A comprehensive description of compute functions can be found in the 
+   C++ implementation docs.
```
Is this a page in the docs we can link to?

jorisvandenbossche reviewed (commented):
Looks good to me!

arw2019 commented on docs/source/python/compute.rst:
```diff
@@ -0,0 +1,55 @@
+
+.. seealso::
+
+   :ref:`C++ compute functions documentation <cpp-compute>`.
```
@jorisvandenbossche Will this work? I'm not 100% sure as I haven't yet compiled the docs to explicitly check

pitrou commented on python/pyarrow/_compute.pyx:
```diff
@@ -669,3 +669,13 @@ cdef class StrptimeOptions(FunctionOptions):
+
+cdef class VarianceOptions(FunctionOptions):
+    cdef:
+        unique_ptr[CVarianceOptions] variance_options
```
I'm not sure why `unique_ptr` is used in some of the other classes, but it shouldn't be necessary. See `MinMaxOptions` for an example without.

arw2019 replied:
In `PartitionNthOptions` we need it because it doesn't have a null constructor so it can't be allocated on the stack. That's not the case for `VarianceOptions`, though, so I'll take a look

arw2019 replied:
I've rewritten `VarianceOptions` without `unique_ptr`.

I think that with the other three options classes exposed in this PR (`SetLookupOptions`, `PartitionNthOptions` and `StrptimeOptions`) it's necessary to use `unique_ptr` or, alternatively, we would need to define default constructors

kszucs commented:
@github-actions crossbow submit test-ubuntu-18.04-docs

pitrou reviewed (approved):
+1. Will merge after a couple changes. Thank you @arw2019 !

pitrou commented:
CI failures are unrelated.

arw2019 commented:
Thanks @pitrou @jorisvandenbossche for reviewing!
